source/extensions/transport_sockets/tls/context_impl.cc

```diff
@@ -52,6 +54,92 @@ bool cbsContainsU16(CBS& cbs, uint16_t n) {
   return false;
 }
 
+int sslTlsContextIndex() {
+  CONSTRUCT_ON_FIRST_USE(int, []() -> int {
+    int ssl_context_index = SSL_CTX_get_ex_new_index(0, nullptr, nullptr, nullptr, nullptr);
+    RELEASE_ASSERT(ssl_context_index >= 0, "");
+    return ssl_context_index;
+  }());
+}
+
+bool seekToSubject(CRYPTO_BUFFER* cert, CBS* tbs_certificate) {
+  CBS der;
+  CRYPTO_BUFFER_init_CBS(cert, &der);
+  CBS certificate, opt_version, serial, signature, issuer, validity;
+  // From RFC 5280, section 4.1
+  //
+  // Certificate  ::=  SEQUENCE  {
+  //      tbsCertificate       TBSCertificate,
+  //      signatureAlgorithm   AlgorithmIdentifier,
+  //      signatureValue       BIT STRING  }
+  //
+  // TBSCertificate  ::=  SEQUENCE  {
+  //      version         [0]  EXPLICIT Version DEFAULT v1,
+  //      serialNumber         CertificateSerialNumber,
+  //      signature            AlgorithmIdentifier,
+  //      issuer               Name,
+  //      validity             Validity,
+  //      subject              Name,
+  //      subjectPublicKeyInfo SubjectPublicKeyInfo,
+  //      ...
+  if (!CBS_get_asn1(&der, &certificate, CBS_ASN1_SEQUENCE) ||
```

Comment:
Please add some test coverage for this. At least the case of of trailing junk after cert, and an invalid cert (either complete junk, or a cert that's in some way invalid (missing a required field, etc)). I want to make sure that the code is exercised a bit since the error path is reachable by client-controlled data.